botocore/compress.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import gzip
+import io
+import logging
+
+from botocore.compat import urlencode
+from botocore.utils import determine_content_length
+
+logger = logging.getLogger(__name__)
+
+
+def maybe_compress_request(config, request_dict, operation_model):
+    """Attempt to compress the request body using the modeled encodings."""
+    if _should_compress_request(config, request_dict, operation_model):
+        for encoding in operation_model.request_compression['encodings']:
+            encoder = COMPRESSION_MAPPING.get(encoding)
+            if encoder is not None:
+                logger.debug('Compressing request with %s encoding.', encoding)
+                request_dict['body'] = encoder(request_dict['body'])
+                _set_compression_header(request_dict['headers'], encoding)
+                return
+            else:
+                logger.debug('Unsupported compression encoding: %s', encoding)
+
+
+def _should_compress_request(config, request_dict, operation_model):
+    if (
+        config.disable_request_compression is not True
+        and config.signature_version != 'v2'
+        and operation_model.request_compression is not None
+    ):
+        # Requests with streaming input are compressed regardless of
+        # `request_min_compression_size_bytes` if they don't contain the
+        # `requiresLength` trait.
+        if operation_model.has_streaming_input:
+            return (
+                'requiresLength'
+                not in operation_model.get_streaming_input().metadata
+            )
+        else:
+            # If the request body is a dictionary, we need to encode it to get the
+            # length. If compression should occur, the encoded body will be saved
+            # to the request dictionary so it doesn't need to be encoded again.
+            body = request_dict['body']
+            if isinstance(body, dict):
+                body = urlencode(body, doseq=True, encoding='utf-8')
+            config_min = config.request_min_compression_size_bytes
+            should_compress = config_min <= _get_body_size(body)
+            if should_compress:
+                request_dict['body'] = body
+            return should_compress
```

Comment:
```suggestion
        if not _attempt_type_normalization(request_dict):
            body_type = type(request_dict['body'])
            logger.debug("Body type {body_type} does not support compression.")
            return False
        if operation_model.has_streaming_input:
            streaming_metadata = operation_model.get_streaming_input().metadata
            return 'requiresLength' not in streaming_metadata

        body_size = _get_body_size(request_dict['body'])
        should_compress = config.request_min_compression_size_bytes <= body_size
        return should_compress

def _attempt_type_normalization(request_dict):
    """Attempt to coerce request body into format for compression."""
    if isinstance(request_dict['body'], dict):
        body = urlencode(request_dict['body'], doseq=True, encoding='utf-8')
        request_dict['body'] = body
    is_supported_type = isinstance(request_dict['body'], [str, bytes, bytesarray])
    return is_supported_type or hasattr(body, 'read')
```

Reply:
Thanks so much @nateprewitt! Taking a look at this now, but I think broad strokes this proposal makes sense.

Reply:
The only functional difference I made is doing the entire encoding (i.e. string to bytes) when coercing the dict into a compressible type. I also changed the name of the function to `_is_compressible_type` since that more closely reflects what the function is returning, while adding a more specific comment instead of a doc string.